### PR TITLE
Proposal: Caching GitHub requests behind API GW

### PR DESCRIPTION
**Pull request by RLRabinowitz**
An alternative solution for https://github.com/opentofu/registry/issues/97
Use API GW with caching to help prevent reaching GitHub throttling errors

### CODE REVIEW

**Comment by Yantrio** [comment]
Overall I love this approach and I think its quite powerful!

As we are introducing so many layers of caching, I think we need to sit down and write out a draft for the timings of what we would like it to be, to make sure that we always fetch fresh data.

For example.

If the API Gateway cache for the api lambda is 1hr
Dynamodb caching for versions should be less than that (55 mins)
And github caching should be less than that again (50 mins?)

This ensures that we have cascading sets of fresh data always coming into the system.

**Comment by rumenvasilev** [comment]
Cross posting for history:

Unless I’m missing something, you can’t cache GraphQL responses, because they are responses to POST requests.
More here: https://www.apollographql.com/blog/backend/caching/graphql-caching-the-elephant-in-the-room/

---
EDIT: Actually you can cache (as @roni-frantchi corrected me on slack), the issue though is when you have multiple customers to the GQL endpoint, then minimal variations in the query would result in a totally new cache key - hitting the cache size and exhausting resources (in AWS that's elastic, so in the end just more money).

**Comment by RLRabinowitz** [comment]
> If the API Gateway cache for the api lambda is 1hr
Dynamodb caching for versions should be less than that (55 mins)
And github caching should be less than that again (50 mins?)

> This ensures that we have cascading sets of fresh data always coming into the system.

So this is not 100% fullproof, since the API GW cache for lambda is per API, but multiple lambdas can call the same GH API (Specifically - All endpoints call the `Get Repository` API, and multiple endpoints will list the releases of a repository). But, with the regular usages of `opentofu` it should always be OK, and the DDB cache + GH API GW cache should be aligned. But with calling the registry API explicitly those caches could not be properly aligned

However, the cost of those caches not being aligned properly is very low. It just increases the time until the registry notices new releases

File: apigw.tf [hunk 1]
```diff
@@ -248,6 +272,66 @@ resource "aws_api_gateway_integration" "metadata_integration" {
...
 
+resource "aws_api_gateway_method" "github_rest_method" {
+  rest_api_id   = aws_api_gateway_rest_api.api.id
+  resource_id   = aws_api_gateway_resource.github_rest_proxy.id
+  http_method   = "GET"
+  authorization = "NONE"
+
+  request_parameters = {
+    "method.request.path.proxy" = true,
+    "method.request.header.Authorization"    = true
```
Comment (cube2222) [inline apigw.tf]: Nit: I believe this is not `terraform fmt`'d

**Review by cube2222 — APPROVED** [review]

**Review by tomasmik — APPROVED** [review]
LGTM but I would suggest someone else to also take a good look at this.

File: src/internal/github/client.go [hunk 2]
```diff
@@ -17,9 +20,11 @@ func getGithubOauth2Client(token string) *http.Client {
 }
 
 func NewManagedGithubClient(token string) *github.Client {
-	return github.NewClient(getGithubOauth2Client(token))
+	client := github.NewClient(getGithubOauth2Client(token))
+	client.BaseURL, _ = url.Parse(fmt.Sprintf("https://%s/github/rest/", os.Getenv("GITHUB_API_GW_URL")))
```
Comment (mbialon) [inline src/internal/github/client.go:24]: Nit: It would be clearer if the GitHub host was an argument.

File: src/internal/github/client.go [hunk 3]
```diff
@@ -17,9 +20,11 @@ func getGithubOauth2Client(token string) *http.Client {
...
+	return client
 }
 
 func NewRawGithubv4Client(token string) *githubv4.Client {
-	return githubv4.NewClient(getGithubOauth2Client(token))
+	return githubv4.NewEnterpriseClient(fmt.Sprintf("https://%s/github/graphql/", os.Getenv("GITHUB_API_GW_URL")), getGithubOauth2Client(token))
```
Comment (mbialon) [inline src/internal/github/client.go:29]: Nit: It would be clearer if the GitHub host was an argument.

File: apigw.tf [hunk 4]
```diff
@@ -390,6 +480,42 @@ resource "aws_api_gateway_method_settings" "well_known_method_settings" {
...
+  # This encodes `/` as `~1` to provide the correct path for the method
+  method_path = "~1github~1graphql~1{proxy}/POST"
+
+  settings {
+    metrics_enabled                         = true
+    logging_level                           = "INFO"
+    data_trace_enabled                      = true
+    caching_enabled                         = true
+    // 50 minutes to keep it consistent with the other caching layers' TTL
+    cache_ttl_in_seconds                    = (50*60)
```
Comment (mbialon) [inline apigw.tf]: @Yantrio's caching PR assumes 55 minutes. Should this be the same?

File: apigw.tf [hunk 5]
```diff
@@ -390,6 +480,42 @@ resource "aws_api_gateway_method_settings" "well_known_method_settings" {
...
+  # This encodes `/` as `~1` to provide the correct path for the method
+  method_path = "~1github~1rest~1/GET"
+
+  settings {
+    metrics_enabled                         = true
+    logging_level                           = "INFO"
+    data_trace_enabled                      = true
+    caching_enabled                         = true
+    // 50 minutes to keep it consistent with the other caching layers' TTL
+    cache_ttl_in_seconds                    = (50*60)
```
Comment (mbialon) [inline apigw.tf]: @Yantrio's caching PR assumes 55 minutes. Should this be the same?

**Review by mbialon — APPROVED** [review]

File: apigw.tf (hunk 4 above)
Comment (RLRabinowitz) [reply apigw.tf]: Due to @Yantrio 's caching PR, this cache needs to be a bit smaller than 55 minutes, to keep the cache mostly consistent with the DynamoDB cache

File: apigw.tf (hunk 5 above)
Comment (RLRabinowitz) [reply apigw.tf]: Due to @Yantrio 's caching PR, this cache needs to be a bit smaller than 55 minutes, to keep the cache mostly consistent with the DynamoDB cache

File: src/internal/github/client.go (hunk 2 above)
Comment (RLRabinowitz) [reply src/internal/github/client.go:24]: I understand this approach, but I imagine we'd want this host to be the default one, so in the end `github/client.go` would still be aware of this host

For now, I would keep this as-is

File: src/internal/github/client.go (hunk 3 above)
Comment (RLRabinowitz) [reply src/internal/github/client.go:29]: I understand this approach, but I imagine we'd want this host to be the default one, so in the end `github/client.go` would still be aware of this host

For now, I would keep this as-is

File: apigw.tf (hunk 1 above)
Comment (RLRabinowitz) [reply apigw.tf]: Ran `terraform fmt`
https://github.com/opentofu/registry/pull/105/commits/8b530e1d304af34b4dbbef9528f6f38b99fff782

**Pull request merged**